42 Add release standards on CONTRIBUTING (#59)
cyberark/community: Conjur/CONTRIBUTING.md

## [0.0.1] - 2020-01-02
### Added
- Added initial functionality foo ([org/repo#123](https://github.com/org/repo/issues/123))

[Unreleased]: https://github.com/org/repo/compare/v0.2.0...HEAD
[0.0.2]: https://github.com/org/repo/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/org/repo/releases/tag/v0.0.1
```

### Changelog Content

While the format of the CHANGELOG is important for machine readability of changes, the actual
content of the included changelog entries is even more critical as it is used by end-users.
Because of this, particular care must be dedicated to the choice of inclusion, style, and
wording of these entries.

This list should give you an idea of what _should_ be included in a changelog:

- Any user-facing change (including additions, changes, and removals of features)
- Changes to the public API endpoint contracts
- Performance improvements that are signifficant enough to be of value to the end-user
- Security fixes
- Changes to supported environments

Some things that **_should not_** be included in the changelog:

- Developer-facing changes (e.g.: test additions, development script changes, etc)

> - Refactorings (unless they are signifficant enough to have risks of new bugs)

Reviewer: @sgnn7 this is still outstanding

- Added FIPS compatibility to the agent ([org/repo#123](#123))
...
```

**Good entries**:
```markdown
...
- Added FIPS compatibility to the agent ([org/repo#123](https://github.com/org/repo/issues/123))
- Removed `-n` flag from the accepted CLI flags ([org/repo#234](https://github.com/org/repo/issues/234))
- Improved performance of main API endpoints with change to
  a new router `foo` package dependency
- Fixed CVE-2020-12345 by upgrading `testing` dependency to v3 ([org/repo#345](https://github.com/org/repo/issues/345))
- Fixed overrides handling of `Client` account param [org/repo#456](https://github.com/org/repo/issues/456)
...
```

## Tagging

All projects cyclically get a point where they are ready to package a group of changes as a
check-pointed release which in most cases is started with creating versions through `git tag`.
Git tags create special branches that allow for a reproducible and easy-to-understand
delivery of code that are generally 1:1 mapped with significant feature changes.

In most cases for our projects we use [annotated git tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags)

> with the name matching the `v<semver>` tri-dotted pattern (e.g. `v1.2.3`). By doing this,

Reviewer: worth adding a link to https://semver.org/ here?

we make finding relevant code and automating releases easy and simple. Where possible, we

> also use GPG signing of those tags though this process has not been yet fully designed.

Reviewer: > though this process has not been yet fully designed
in what way? I think this comment could be clearer, or maybe even just drop this and leave as "Where possible, we do this"

To create an annotated tag you have to:
- **Check out the default branch (in most cases this is `master`) and/or the commit that you
  want to tag**. In almost all cases, tags should be created on the default branch for
  automation purposes as a lot of our deliverables get pushed only when the default branch
  has a specific tag. This checkout step is somewhat optional as `git tag` can also take
  in a commit to tag but this procedure is less error-prone vs direct tagging.

  Example of checking out the current upstream main branch history
  ```sh-session
  $ git fetch
  $ git checkout master
  $ git reset --hard origin/master
  ```
- **Verify that the git history looks as expected**. Tools like `git log`, `tig`, `gitk`, and
  others can help you verify that the history is in the expected state.
- **Create the annotated tag**. To create the tag, use the `git tag -a` command. The main tag
  **must** be in `v<semver>` tri-dotted format:
  ```sh-session
  $ # If you don't have a GPG key set up
  $ git tag -a "v1.2.3"

  $ # If you do have a GPG key set up
  $ git tag -s "v1.2.3"

>   $ # Make sure to add a commit header and description when the editor window opens

Reviewer: I always make my message simply `v1.2.3`, and leave the additional description to the github release - wdyt about giving that general instruction here? if you think more detail is required, is there an example header / description that you'd add?

  ```

- **Verify that the tag was created correctly**. You can check this easily with `git describe`:
  ```sh-session
  $ git describe
  v1.2.3
  ```
  If the version returned by `git describe` does not match your created tag, it was not correctly
  created.
- **Push the tag to the upstream repository**. This step makes the tag available to other users
  of the repo:
  ```sh-session
  $ git push origin v1.2.3
  ```

  > Note: Overwriting or removing old tags is extremely discouraged

## GitHub Releases

GitHub releases are the main way that our users consume some of our end products so it is very
important that these are maintained 1:1 along with tags.

> > **Note: Only full releases are considered for inclusion in the Conjur OSS Suite**

Reviewer: ```suggestion
> **Note: Only full (i.e. not draft or pre-release) GitHub releases are considered for inclusion in the Conjur OSS Suite**
```

### Draft Release Creation

Once the [tag](#tagging) has been pushed to upstream repository and you are ready to create an
official release of that tagged version we can go to the GitHub UI to first create a draft release:
- **Navigate to the releases page**. In GitHub, under `<main_repo_url>/releases` you can find
  the list of all relases that are available for the repository. We will click the `Draft a new release`
  in the top right to get us into the "new release" page.
- **Fill out the relevant release info**. Make sure that:
  - `Tag version` is the tag you created for this release.
  - `Release title` matches exactly the tag name (`v<semver>`).

>   - `Describe this release` is populated with changelog entries for this tagged version.

Reviewer: it might be worth adding more detail here - that you should add a header `## Change log` and then copy / paste the section from the CHANGELOG.md file for this version into the release description

Author: Header might not be needed but I'll add the note about copy/pasting.

>   - Any files that you may want included this release are attached. This typically includes the

Reviewer: + NOTICES.txt

    `LICENSE.md` file, `CHANGELOG.md` file, binary executable (if any), SHA256SUM of the
    executable artifacts, etc).
- **Create the draft release**. Check the box `This is a pre-release` and click `Save Draft`
  which will create a draft release that is not yet visible to the public.

Once the draft release is created it's a good time to triple-check the release and ensure that
all of the executable files pass a smoke test for their environments. While this is not strictly
mandatory, it prevents issues with mis-labeled or incompatible binaries.

### Pre-release Publishing

With draft release smoke test completed, we can make a "pre-release" (non-production ready
release) to let the users know that we have a new release but that we are performing a slow
rollout in an effort to ensure low impact if any bugs are found in this new release. This

> stage may last as little or as much as needed and it's at the discretion of the maintainers

Reviewer: ```suggestion
stage may last as little or as much as needed and it's at the discretion of the maintainers,
```

> but should be guided by thinks like: the size of user base, cascading impact for bugs, level

Reviewer: ```suggestion
but should be guided by things like: the size of user base, cascading impact for bugs, level
```

of automated testing in the repo, risk of changes causing unintended problems, and other
concerns.

To make a pre-release:
- **Go to the draft release page**. The draft release can be found under
  `<main_repo_url>/releases` for your repo.
- **Publish the pre-release**. Double-check that the `This is a pre-release` is selected and
  click `Publish release`.

### Release Publishing

After some (project-dependent) period of time of usage by the users of the pre-released
version of the software, the confidence that the software works as expected is much higher

> and we can then turn our pre-release into a full release:

Reviewer: some projects may have specific requirements to go from prerelease to release

for example, we tag prereleases in secretless for new tags bc at the time of tagging, we've run through a comprehensive test suite and have confidence to make a release but haven't done stability / performance tests to say it's production ready yet

once we've reviewed the changes and either determined they're not going to impact performance or we've validated the performance via testing so that we have some assurance the release meets our [stability requirements](https://github.com/cyberark/secretless-broker/blob/master/README.md#stable-release-definition), _then_ we change the pre-release to a release

which is a very long winded way of saying that it's not necessarily "after a period of time", but it's a project-specific definition of production readiness requirements that must be met to turn a pre-release into a release

To create an annotated tag you have to:
- **Check out the default branch (in most cases this is `master`) and/or the commit that you
  want to tag**. In almost all cases, tags should be created on the default branch for
  automation purposes as a lot of our deliverables get pushed only when the default branch
  has a specific tag. This checkout step is somewhat optional as `git tag` can also take
  in a commit to tag but this procedure is less error-prone vs direct tagging.

  Example of checking out the current upstream main branch history
  ```sh-session
  $ git fetch
  $ git checkout master
  $ git reset --hard origin/master
  ```
- **Verify that the git history looks as expected**. Tools like `git log`, `tig`, `gitk`, and
  others can help you verify that the history is in the expected state.
- **Create the annotated tag**. To create the tag, use the `git tag -a` command. The main tag
  **must** be in `v<semver>` tri-dotted format:
  ```sh-session
  $ # If you don't have a GPG key set up
  $ git tag -a "v1.2.3" -m "v1.2.3"

  $ # If you do have a GPG key set up
  $ git tag -s "v1.2.3" -m "v1.2.3"
  ```

> 

Reviewer: I think we should maybe mention verifying that the CHANGELOG looks correct and if the dependencies were updated that the NOTICES.txt is also up to date

### Pre-release Publishing

With draft release smoke test completed, we can make a "pre-release" (non-production ready
release) to let the users know that we have a new release but that we are performing a slow
rollout in an effort to ensure low impact if any bugs are found in this new release. This
stage may last as little or as much as needed and it's at the discretion of the maintainers,
but should be guided by things like: the size of user base, cascading impact for bugs, level
of automated testing in the repo, risk of changes causing unintended problems, and other
concerns.

To make a pre-release:
- **Go to the draft release page**. The draft release can be found under
  `<main_repo_url>/releases` for your repo.
- **Publish the pre-release**. Double-check that the `This is a pre-release` is selected and
  click `Publish release`.

### Release Publishing

After some (project-dependent) assurance that the release meets our stability requirements by
the developers and users of the pre-released version of the software, the confidence that the
software works as expected is much higher and we can then be confident that the software works

> as expected. When this stage is reached, we can turn our pre-release into a full release:

Reviewer: this is redundant now - "works as expected" x 2

- Removed `-n` flag from the accepted CLI flags ([org/repo#234](https://github.com/org/repo/issues/234))
- Improved performance of main API endpoints with change to
  a new router `foo` package dependency
- Fixed CVE-2020-12345 by upgrading `testing` dependency to v3 ([org/repo#345](https://github.com/org/repo/issues/345))
- Fixed overrides handling of `Client` account param [org/repo#456](https://github.com/org/repo/issues/456)
...
```

## Tagging

All projects cyclically get a point where they are ready to package a group of changes as a
check-pointed release which in most cases is started with creating versions through `git tag`.
Git tags create special branches that allow for a reproducible and easy-to-understand
delivery of code that are generally 1:1 mapped with significant feature changes.

In most cases for our projects we use [annotated git tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags)
with the name matching the [`v<semver>` tri-dotted pattern](https://semver.org/) (e.g.
`v1.2.3`). By doing this, we make finding relevant code and automating releases easy and

> simple. We also use GPG signing of those tags where possible but the infrastructure and
> guidelines for signing tags have not been fully worked out on an official level.

Reviewer: ```suggestion
simple. We also use GPG signing of those tags where possible, but this can vary across
projects as our official process for this is still being defined.
```

To create an annotated tag you have to:
- **Check out the default branch (in most cases this is `master`) and/or the commit that you
  want to tag**. In almost all cases, tags should be created on the default branch for
  automation purposes as a lot of our deliverables get pushed only when the default branch
  has a specific tag. This checkout step is somewhat optional as `git tag` can also take
  in a commit to tag but this procedure is less error-prone vs direct tagging.

  Example of checking out the current upstream main branch history
  ```sh-session
  $ git fetch
  $ git checkout master
  $ git reset --hard origin/master
  ```
- **Verify that the git history looks as expected**. Tools like `git log`, `tig`, `gitk`, and
  others can help you verify that the history is in the expected state.

> - **Verify that `NOTICES.txt` looks as expected**. This file should include any info
>   that the project may have to comply with legally, including:

Reviewer: ```suggestion
- **Verify that `NOTICES.txt` looks as expected**. In particular, if the changes in this version
   have modified the project dependencies, this file likely will need an update. In general, this file
   includes info to help us comply with the legal requirements of our open source dependencies,
   including:
```

  - Legally required copyright notices
  - Legally required license texts
  - Legally required upstream repository links
  - Any other required legal information about the project and its dependencies
- **Verify that the `CHANGELOG.md` content looks as expected**. While most repos already
  have tooling for this included in the CI/CD pipeline, it's always a good idea to verify that:

>   - The [changelog guidelines](#changelog-guiedlines) were followed

Reviewer: ```suggestion
  - The [changelog guidelines](#changelog-guidelines) were followed
```

>   - And that the tag you are creating has a matching version section in the CHANGELOG

Reviewer: ```suggestion
  - The tag you are creating has a matching version section in the CHANGELOG.
```
